### PR TITLE
IDforwarding: forward signed id to plugins

### PR DESCRIPTION
**Pull request by kalleep**
**What is this feature?**
We want the ability to forward a singed id token for the user making the request. The signing of tokens happens independent of plugins calls and are cached. For now this feature is protected with a feature flag that requires grafana to run in dev mode.

This pr just adds a client middleware and functionality to DsProxy that will set `X-Grafana-Id` header if we have a generated token for the user.

Something we probably should do is add the ability to configure if the id token should be forwarded to a data source. This can be done in a follow-up pr 

**Which issue(s) does this PR fix?**:
Part of https://github.com/grafana/identity-access-team/issues/310

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/api/pluginproxy/ds_proxy.go [hunk 1]
```diff
@@ -262,6 +262,8 @@ func (proxy *DataSourceProxy) director(req *http.Request) {
 			}
 		}
 	}
+
+	proxyutil.ApplyForwardIDHeader(req, proxy.ctx.SignedInUser)
```
Comment (andresmgot) [inline pkg/api/pluginproxy/ds_proxy.go:266]: just to be extra-safe, can you add the middleware only if the feature flag is enabled?

File: pkg/services/pluginsintegration/pluginsintegration.go [hunk 2]
```diff
@@ -158,6 +158,7 @@ func CreateMiddlewares(cfg *setting.Cfg, oAuthTokenService oauthtoken.OAuthToken
 		clientmiddleware.NewTracingHeaderMiddleware(),
 		clientmiddleware.NewClearAuthHeadersMiddleware(),
 		clientmiddleware.NewOAuthTokenMiddleware(oAuthTokenService),
+		clientmiddleware.NewForwardIDMiddleware(),
```
Comment (andresmgot) [inline pkg/services/pluginsintegration/pluginsintegration.go:161]: same here, only add it if the feature flag is enabled

**Review by andresmgot — COMMENTED** [review]
Code LGTM but I'd suggest to put the new code behind the feature flag rather than relying that the IDToken will be empty

**Review by Jguer — APPROVED** [review]
setter logic seems correct :+1:.

**Comment by kalleep** [comment]
@andresmgot I added feature toggle checks. Also took the liberty to clean up tests setups in `ds_proxy_test.go`.

**Comment by kalleep** [comment]
@Jguer Do you think we should add id forwarding to [plugin-proxy](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/pluginproxy.go)? 

Seems like we are forwarding the [unsigned user identity](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/pluginproxy.go#L157) there

File: pkg/services/pluginsintegration/clientmiddleware/forward_id_middleware.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+			next: next,
+		}
+	})
+}
+
+type ForwardIDMiddleware struct {
+	next plugins.Client
+}
+
+type requestWithHeader interface {
```
Comment (marefr) [inline pkg/services/pluginsintegration/clientmiddleware/forward_id_middleware.go:28]: Nit. Is this really needed. Can't you use the [backend.ForwardHTTPHeaders](https://github.com/grafana/grafana-plugin-sdk-go/blob/4aae405d0a88b4bca46e7e3276a535cb8b3d4236/backend/http_headers.go#L27C6-L27C24) interface?

**Review by marefr — APPROVED** [review]
LGTM with a comment. Agree that plugin proxy probably should forward this as well as earlier discussed.

File: pkg/services/pluginsintegration/clientmiddleware/forward_id_middleware.go (hunk 3 above)
Comment (kalleep) [reply pkg/services/pluginsintegration/clientmiddleware/forward_id_middleware.go]: Fixed in [fef8dd4](https://github.com/grafana/grafana/pull/75651/commits/fef8dd40e00a024170f58606e3adb841ec935938)

**Comment by kalleep** [comment]
Will add it to plugin proxy also :)

**Review by andresmgot — APPROVED** [review]
:+1: LGTM!

**Pull request merged**